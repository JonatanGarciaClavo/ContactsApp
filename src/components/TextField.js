import React, { PropTypes } from 'react';
import TextFieldMUI from 'material-ui/TextField';
import _ from 'lodash';

class TextField extends React.Component {

  constructor(props) {
    super(props);
    this.onChange = this.onChange.bind(this);
    this.onBlur = this.onBlur.bind(this);
    this.state = {
      value: props.value,
    };
    this.cbOnChange = _.debounce(props.onChange, 250);
  }

  componentWillReceiveProps(nextProps) {
    if (this.state.value !== nextProps.value) {
      this.setState({ value: nextProps.value });
    }
  }

  onChange(e) {
    const { name, onChange } = this.props;
    const value = e.target.value;
    this.setState({ value });
    if (onChange) {
      this.cbOnChange(name, value);
    }
  }

  onBlur(e) {
    const { name, onBlur } = this.props;
    if (onBlur) {
      onBlur(name, e.target.value);
    }
  }

  render() {
    const { label, name, placeholder, type, errorText, onEnterKeyDown } = this.props;
    return (
      <TextFieldMUI
        value={this.state.value}
        floatingLabelText={label || name}
        hintText={placeholder}
        type={type || 'text'}
        errorText={errorText}
        onChange={this.onChange}
        onBlur={this.onBlur}
        onEnterKeyDown={onEnterKeyDown}
      />

    );
  }
}

TextField.propTypes = {
  name: PropTypes.string.isRequired,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]).isRequired,
  label: PropTypes.string,
  placeholder: PropTypes.string,
  type: PropTypes.string,
  errorText: PropTypes.string,
  onChange: PropTypes.func,
  onBlur: PropTypes.func,
  onEnterKeyDown: PropTypes.func,
};

export default TextField;
