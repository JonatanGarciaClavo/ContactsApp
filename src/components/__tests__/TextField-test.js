import React from 'react';
import TextField from '../TextField';

describe('Testing TextField component', () => {
  it('renders TextField using Snapshots', () => {
    const wrapper = shallow(
      <TextField
        label="Name"
        placeholder="Name"
        name="name"
        value={''}
        errorText={null}
        onChange={jest.fn}
        onBlur={jest.fn}
        onEnterKeyDown={jest.fn}
      />
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
  it('check change functionality', () => {
    const name = 'name';
    const value = 'Test';
    const onChange = jest.fn();
    const wrapper = shallow(
      <TextField
        label="Name"
        placeholder="Name"
        name={name}
        value={''}
        errorText={null}
        onChange={onChange}
        onBlur={jest.fn}
        onEnterKeyDown={jest.fn}
      />
    );
    wrapper.find('TextField').simulate('change', {
      target: { value },
    });
    expect(wrapper.state().value).toEqual(value);
  });
  it('check change functionality without callback', () => {
    const name = 'name';
    const value = 'Test';
    const wrapper = shallow(
      <TextField
        label="Name"
        placeholder="Name"
        name={name}
        value={''}
        errorText={null}
        onChange={null}
        onBlur={jest.fn}
        onEnterKeyDown={jest.fn}
      />
    );
    wrapper.find('TextField').simulate('change', {
      target: { value },
    });
    expect(wrapper.state().value).toEqual(value);
  });
  it('check blur functionality', () => {
    const name = 'name';
    const value = 'Test';
    const onBlur = jest.fn();
    const wrapper = shallow(
      <TextField
        label="Name"
        placeholder="Name"
        name={name}
        value={value}
        errorText={null}
        onChange={jest.fn}
        onBlur={onBlur}
        onEnterKeyDown={jest.fn}
      />
    );
    wrapper.find('TextField').simulate('blur', {
      target: { value },
    });
    expect(onBlur).toBeCalledWith(name, value);
  });
  it('check blur functionality without callback', () => {
    const value = 'Test';
    const wrapper = shallow(
      <TextField
        label="Name"
        placeholder="Name"
        name={name}
        value={''}
        errorText={null}
        onChange={jest.fn}
        onBlur={null}
        onEnterKeyDown={jest.fn}
      />
    );
    wrapper.find('TextField').simulate('blur', {
      target: { value },
    });
  });
  it('check change value when recive new value prop', () => {
    const value = 'Test';
    const wrapper = shallow(
      <TextField
        placeholder="Name"
        name={name}
        value={''}
        errorText={null}
        onChange={jest.fn}
        onBlur={jest.fn}
        onEnterKeyDown={jest.fn}
      />
    );
    expect(wrapper.state().value).toEqual('');
    wrapper.setProps({ value });
    expect(wrapper.state().value).toEqual(value);
  });
  it('check change value when recive same value prop', () => {
    const value = 'Test';
    const wrapper = shallow(
      <TextField
        placeholder="Name"
        name={name}
        value={value}
        errorText={null}
        onChange={jest.fn}
        onBlur={jest.fn}
        onEnterKeyDown={jest.fn}
      />
    );
    expect(wrapper.state().value).toEqual(value);
    wrapper.setProps({ value });
    expect(wrapper.state().value).toEqual(value);
  });
});
