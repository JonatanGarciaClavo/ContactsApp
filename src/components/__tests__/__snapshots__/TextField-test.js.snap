exports[`Testing TextField component renders TextField using Snapshots 1`] = `
<TextField
  disabled={false}
  errorText={null}
  floatingLabelFixed={false}
  floatingLabelText="Name"
  fullWidth={false}
  hintText="Name"
  multiLine={false}
  onBlur={[Function]}
  onChange={[Function]}
  rows={1}
  type="text"
  underlineShow={true}
  value="" />
`;
